### PR TITLE
Add on-host tests and results upload

### PR DESCRIPTION
**Pull request by imindich**
b/371590965

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -126,6 +138,7 @@ jobs:
         config: [devel, debug, qa, gold]
     container: ${{ needs.docker-build-image.outputs.docker_tag }}
     env:
+      TEST_ARTIFACTS_KEY: ${{ matrix.name }}_test_artifacts
```
Comment (isarkis) [inline .github/workflows/main.yaml:141]: Let's change this to include platform name, otherwise test tar archive name collides between linux and linux-chrobalt platforms.  Something like so:

TEST_ARTIFACTS_KEY: ${{ matrix.platform }}_${{ matrix.name }}_test_artifacts

File: .github/actions/on_host_tests/action.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      shell: bash
+      env:
+        GTEST_OUTPUT_ARG: ${{ inputs.results_dir }}
+        XVFB_SERVER_ARGS: "-screen 0 1920x1080x24i +render +extension GLX -noreset"
+      run: |
+        set -x
+        cd ${GITHUB_WORKSPACE}/unit_test
+        for test_binary in $(ls -d out/**/*tests); do
+          # TODO(b/372303096): Need dedicated xvfb runner.
+          xvfb-run -a --server-args=${XVFB_SERVER_ARGS} ./$test_binary --gtest_output="xml:${RESULTS_DIR}/${test_binary}_testoutput.xml"
```
Comment (isarkis) [inline .github/actions/on_host_tests/action.yaml:35]: let's use an absolute path (i.e. /usr/bin/xvfb-run)

File: .github/actions/on_host_tests/action.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        tar xvf ${GITHUB_WORKSPACE}/test_artifacts.tar.xz
+    - name: Run Tests
+      id: run-tests
+      shell: bash
+      env:
+        GTEST_OUTPUT_ARG: ${{ inputs.results_dir }}
+        XVFB_SERVER_ARGS: "-screen 0 1920x1080x24i +render +extension GLX -noreset"
+      run: |
+        set -x
+        cd ${GITHUB_WORKSPACE}/unit_test
```
Comment (isarkis) [inline .github/actions/on_host_tests/action.yaml:32]: can you run `env` here, it will show environmental variables including PATH?

**Review by isarkis — DISMISSED** [review]

File: .github/actions/on_host_tests/action.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        pwd
+        ls
+        ls ${GITHUB_WORKSPACE}
+        echo "Attempting to run tests..."
+
+        cd ${GITHUB_WORKSPACE}/unit_test
+        for test_binary in $(ls -d out/**/*tests); do
+          # TODO(b/372303096): Need dedicated xvfb runner.
+          echo "Running tests for suite: ${test_binary}"
+          /usr/bin/xvfb-run -a --server-args="${XVFB_SERVER_ARGS}" ./$test_binary --gtest_filter=`tr '\n' ':' < ${GITHUB_WORKSPACE}/src/cobalt/testing/failing_tests.filter` --gtest_output="xml:${RESULTS_DIR}/${test_binary}_testoutput.xml" || true
```
Comment (kaidokert) [inline .github/actions/on_host_tests/action.yaml:43]: You are using `tr` here to to take the text file and pass it to args - neat trick.

I'd recommend try a bit with `jq` - json query tool. Would allow making filter files json - easier to check for correct syntax and also otherwise easily machine readable.

E.g : `jq -r '.filters | join(":")' file.json

And filters.json could look like

```
{
  "filters": [
    "-foo", "-bar", "baz"
   ]
}
```

Comment (kaidokert) [reply .github/actions/on_host_tests/action.yaml:43]: I think from your current test what was missing is that you need to enclose the filters in double quotes for the shell, e.g

```
--gtest_filter="thefilterspec goes here"
```

Because otherwise characters like `-` and `*` will get interpreted as shell tokens, not filterspec

File: cobalt/testing/failing_tests_filter.json [hunk 5]
```diff
@@ -0,0 +1,106 @@
+{
+  "filters": [
+    "-FileUtilTest.ChangeDirectoryPermissionsAndEnumerate",
```
Comment (kaidokert) [inline cobalt/testing/failing_tests_filter.json:3]: Quick nit: Can you sort this file alphabetically please ? ( in VScode, select lines, cmd+p "sort" )

File: cobalt/testing/failing_tests_filter.json [hunk 6]
```diff
@@ -0,0 +1,106 @@
+{
+  "filters": [
```
Comment (kaidokert) [inline cobalt/testing/failing_tests_filter.json:2]: Also if you rename this key from "filters" to "excluded_tests" perhaps it makes more obvious sense, and you can omit the `-` sign from the filter names and add that in your jquery expression. Would make it less visually noisy and easier to comprehend what's going on in the file

File: cobalt/testing/failing_tests_filter.json (hunk 5 above)
Comment (imindich) [reply cobalt/testing/failing_tests_filter.json]: Done.

File: cobalt/testing/failing_tests_filter.json (hunk 6 above)
Comment (imindich) [reply cobalt/testing/failing_tests_filter.json]: Thanks, done.

File: .github/actions/build/action.yaml [hunk 7]
```diff
@@ -33,30 +36,41 @@ runs:
...
-      - name: sccache check
+      - name: Dump sccache stats
         run: sccache -s
         shell: bash
-      - name: find apks
-        run: find src/out -name \*.apk -print
-        shell: bash
       - name: Archive Android APKs
-        if: startsWith(${{matrix.platform}}, 'android')
+        if: startsWith(${{matrix.platform}}, 'android') && matrix.config == 'qa'
```
Comment (imindich) [inline .github/actions/build/action.yaml:56]: @oxve Is this change necessary?

Comment (oxve) [reply .github/actions/build/action.yaml:56]: I don't remember why it is there. Probably not.

File: .github/actions/build/action.yaml [hunk 8]
```diff
@@ -60,3 +60,17 @@ runs:
           path: |
             src/out/${{ matrix.platform }}_qa/apks/*.apk
             src/out/${{ matrix.platform }}_qa/*_apk/*.apk
+      - name: Create Test Files Archive
+        if: startsWith(matrix.platform, 'linux') && matrix.config == 'devel'
+        run: |
+          set -x
+          cd src
+          tar cvf - out/${{ matrix.platform }}_${{ matrix.config }}/*tests out/${{ matrix.platform }}_${{ matrix.config }}/*.so out/${{ matrix.platform }}_${{ matrix.config }}/starboard/*.so | xz -T0 -1 -z - > test_artifacts.tar.xz
```
Comment (oxve) [inline .github/actions/build/action.yaml:68]: >out/${{ matrix.platform }}_${{ matrix.config }}/starboard/*.so

These are not the file you are looking for. The filename is `libstarboard.so.17` so the glob fails to match.

Change the starboard glob to `*.so.*`.

Comment (imindich) [reply .github/actions/build/action.yaml]: Done.

File: .github/workflows/main.yaml [hunk 9]
```diff
@@ -147,3 +160,48 @@ jobs:
...
+        platform: ${{ fromJson(needs.initialize.outputs.platforms) }}
+        include: ${{ fromJson(needs.initialize.outputs.includes) }}
+        config: [devel]
+    container: ${{ needs.docker-build-image.outputs.docker_tag }}
+    env:
+      TMPDIR: /__w/_temp
+      TEST_RESULTS_DIR: ${{ matrix.name }}_test_results
+      TEST_ARTIFACTS_KEY: ${{ matrix.platform }}_${{ matrix.name }}_test_artifacts
+      IS_POSTSUBMIT: ${{ github.event_name == 'schedule' || github.event_name == 'push' }}
+      TESTING_ENABLED: ${{ needs.initialize.outputs.test_on_host == 'true' || needs.initialize.outputs.test_on_device == 'true' }}
```
Comment (oxve) [inline .github/workflows/main.yaml:183]: `TESTING_ENABLED` is unused. Please remove.

File: .github/workflows/main.yaml [hunk 10]
```diff
@@ -147,3 +160,48 @@ jobs:
...
+      matrix:
+        platform: ${{ fromJson(needs.initialize.outputs.platforms) }}
+        include: ${{ fromJson(needs.initialize.outputs.includes) }}
+        config: [devel]
+    container: ${{ needs.docker-build-image.outputs.docker_tag }}
+    env:
+      TMPDIR: /__w/_temp
+      TEST_RESULTS_DIR: ${{ matrix.name }}_test_results
+      TEST_ARTIFACTS_KEY: ${{ matrix.platform }}_${{ matrix.name }}_test_artifacts
+      IS_POSTSUBMIT: ${{ github.event_name == 'schedule' || github.event_name == 'push' }}
```
Comment (oxve) [inline .github/workflows/main.yaml:182]: `IS_POSTSUBMIT` is unused. Please remove.

File: .github/actions/on_host_tests/action.yaml [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      id: run-tests
+      shell: bash
+      env:
+        RESULTS_DIR: ${{ inputs.results_dir }}
+        XVFB_SERVER_ARGS: "-screen 0 1920x1080x24i +render +extension GLX -noreset"
+      run: |
+        set -x
+        env
+        pwd
+        ls
```
[the remark this answers is not in the document]
Comment (imindich) [reply .github/actions/on_host_tests/action.yaml]: Done.

File: .github/workflows/main.yaml (hunk 9 above)
Comment (imindich) [reply .github/workflows/main.yaml]: Done.

File: .github/workflows/main.yaml (hunk 10 above)
Comment (imindich) [reply .github/workflows/main.yaml]: Done.

File: cobalt/testing/linux/my_output2.xml [hunk 12]
```diff
@@ -0,0 +1,123 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kaidokert) [inline cobalt/testing/linux/my_output2.xml:1]: This file probably not needed :)

Comment (imindich) [reply cobalt/testing/linux/my_output2.xml]: Removed, thanks

File: build/util/LASTCHANGE.dummy [hunk 13]
```diff
@@ -1 +1,2 @@
-LASTCHANGE=0000000000000000000000000000000000000000-0000000000000000000000000000000000000000
```
Comment (oxve) [inline build/util/LASTCHANGE.dummy:1]: Is this change intentional?

File: .github/actions/build/action.yaml [hunk 14]
```diff
@@ -60,3 +60,17 @@ runs:
...
+          cd src
+          tar cvf - out/${{ matrix.platform }}_${{ matrix.config }}/*tests out/${{ matrix.platform }}_${{ matrix.config }}/*.so out/${{ matrix.platform }}_${{ matrix.config }}/*.so.* out/${{ matrix.platform }}_${{ matrix.config }}/starboard/*.so.* | xz -T0 -1 -z - > test_artifacts.tar.xz
+        shell: bash
+      - name: Upload Test artifacts
+        if: startsWith(matrix.platform, 'linux') && matrix.config == 'devel'
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ inputs.test_artifacts_key }}
+          path: src/test_artifacts.tar.xz
+          retention-days: 3
```
Comment (oxve) [inline .github/actions/build/action.yaml:76]: `IS_POSTSUBMIT` and `TESTING_ENABLED` was used for uploading to datadog. Can you add that step back in and guard it with those two env variables? Also add `continue-on-error: true` for this step so it doesn't affect the job status.

It should work once merged and is an important piece of this that would be nice to have from day one.

File: .github/actions/on_host_tests/action.yaml [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        # Update path to find missing library linkages.
+        export LD_LIBRARY_PATH="${GITHUB_WORKSPACE}"/unit_test/out/${{ matrix.platform }}_${{ matrix.config }}/starboard:"${GITHUB_WORKSPACE}"/unit_test/out/${{ matrix.platform }}_${{ matrix.config }}
+
+        cd ${GITHUB_WORKSPACE}/unit_test
+        for test_binary in $(ls -d out/**/*tests); do
+          # TODO(b/372303096): Need dedicated xvfb runner.
+          echo "Running tests for suite: ${test_binary}"
+
+          test_filter="*"
+          if [ -f ${GITHUB_WORKSPACE}/src/cobalt/testing/${{ inputs.platform }}/${test_binary##*/}_filter.json ]; then
```
Comment (isarkis) [inline .github/actions/on_host_tests/action.yaml:46]: Does this mean every unit test apk should have a corresponding json file? What if there is nothing to filter?

Comment (imindich) [reply .github/actions/on_host_tests/action.yaml:46]: No, the `if` check here checks for the existence of a json filter file. If it doesn't exist, the default test filter `"*"` includes all tests.

File: build/util/LASTCHANGE.dummy (hunk 13 above)
Comment (imindich) [reply build/util/LASTCHANGE.dummy:1]: Yes, this is a hotfix recommended in b/348690792 to get net_unittests to run. Will remove if it doesn't end up changing anything.

Comment (imindich) [reply build/util/LASTCHANGE.dummy]: Removed

**Review by kaidokert — APPROVED** [review]
Hey - i've looked at this a couple times, i see green tests. I'm okay with this landing.

Pleas make 100% sure that you use "Squash Merge" here because we absolutely do not want the history of all the commits here landed :)

**Review by isarkis — APPROVED** [review]
